middleware/context_timeout.go

```diff
@@ -0,0 +1,92 @@
+package middleware
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/labstack/echo/v4"
+)
+
+// Simply return or wrap the error that context passed method returns.
+// Note: Injects ContextWithTimeout to c.Request().Context(). Does not work with kong running processes without context.
+//
+// e.GET("/", func(c echo.Context) error {
+// 	sleepTime := time.Duration(2 * time.Second)
+//
+// 	if err := sleepWithContext(c.Request().Context(), sleepTime); err != nil {
+// 		return fmt.Errorf("%w: execution error", err)
+// 	}
+//
+// 	return c.String(http.StatusOK, "Hello, World!")
+// })
+//
+
+// ContextTimeoutConfig defines the config for ContextTimeout middleware.
+type ContextTimeoutConfig struct {
+	// Skipper defines a function to skip middleware.
+	Skipper Skipper
+
+	// ErrorHandler is a function when error aries in middeware execution.
+	ErrorHandler func(err error, c echo.Context) error
+
+	// Timeout configures a timeout for the middleware, defaults to 0 for no timeout
+	Timeout time.Duration
+}
+
+var (
+	// DefaultContextTimeoutErrorHandler is default error handler of ContextTimeout middleware.
+	DefaultContextTimeoutErrorHandler = func(err error, c echo.Context) error {
+		if err != nil {
+			if errors.Is(err, context.DeadlineExceeded) {
+				return echo.ErrServiceUnavailable
+			}
+			return err
+		}
+		return nil
+	}
+
+	// DefaultContextTimeoutConfig is the default ContextTimeoutConfig middleware config.
+	DefaultContextTimeoutConfig = ContextTimeoutConfig{
+		Skipper:      DefaultSkipper,
+		Timeout:      0,
+		ErrorHandler: DefaultContextTimeoutErrorHandler,
+	}
+)
+
+// ContextTimeout returns a middleware which returns error (503 Service Unavailable error) to client
+// when underlying method returns context.DeadlineExceeded error.
+func ContextTimeout() echo.MiddlewareFunc {
+	return ContextTimeoutWithConfig(DefaultContextTimeoutConfig)
+}
+
+// ContextTimeoutWithConfig returns a Timeout middleware with config.
+func ContextTimeoutWithConfig(config ContextTimeoutConfig) echo.MiddlewareFunc {
+	return config.ToMiddleware()
+}
+
+// ToMiddleware converts Config to middleware.
+func (config ContextTimeoutConfig) ToMiddleware() echo.MiddlewareFunc {
```

Comment:
I think this method can be changed to

```go
func ContextTimeout(timeout time.Duration) echo.MiddlewareFunc {
	return ContextTimeoutWithConfig(ContextTimeoutConfig{Timeout: timeout})
}

func ContextTimeoutWithConfig(config ContextTimeoutConfig) echo.MiddlewareFunc {
	mw, err := config.ToMiddleware()
	if err != nil {
		panic(err)
	}
	return mw
}

func (config ContextTimeoutConfig) ToMiddleware() (echo.MiddlewareFunc, error) {
	if config.Timeout == 0 {
		return nil, errors.New("timeout must be set")
	}
	if config.Skipper == nil {
		config.Skipper = DefaultSkipper
	}
	if config.ErrorHandler == nil {
		config.ErrorHandler = func(err error, c echo.Context) error {
			if err != nil && errors.Is(err, context.DeadlineExceeded) {
				return echo.ErrServiceUnavailable.WithInternal(err)
			}
			return err
		}
	}

	return func(next echo.HandlerFunc) echo.HandlerFunc {
		return func(c echo.Context) error {
			if config.Skipper(c) {
				return next(c)
			}

			timeoutContext, cancel := context.WithTimeout(c.Request().Context(), config.Timeout)
			defer cancel()

			c.SetRequest(c.Request().WithContext(timeoutContext))

			if err := next(c); err != nil {
				return config.ErrorHandler(err, c)
			}
			return nil
		}
	}, nil
}
```

Reply:
Thanks for your suggestion.

I have some doubts on panic the application when the timeout value equals 0. Anyone can inject the timeout duration value via configuration and may disable it by simply setting 0.  

What do you think? Is it a valid scenario? or should we ignore that case?

Reply:
I think it would be better for force user to handle 0 timeout for themselves. Reasoning here is that "Timeout" middleware is about certain amount of time it takes to timeout. This is essential about this middleware. It is like allowing creating JWT middleware without secret and making it do skip checks in that case.   (nb: I aware that other timeout middleware allows this but this is testament that we make sometimes mistakes).  Also it would be very slightly more performant if timeout is 0 that in that case that middleware would not be executed at all.

I would force people to use explicit checks like below and not  rely on magic no-op logic.
```go
	if conf.timeout != 0 {
		e.Use(middleware.ContextTimeout(20 * time.Second))
	}
```